lib/companies/companies.ex

```diff
@@ -100,6 +100,30 @@ defmodule Companies.Companies do
     |> Repo.get!(id)
   end
 
+  @doc """
+  Gets a single company.
+
+  Raises `Ecto.NoResultsError` if the Company does not exist.
+
+  ## Examples
+
+  iex> get!("Valid slug")
+  %Company{}
+
+  iex> get!("Invalid slug")
+  ** (Ecto.NoResultsError)
+
+  """
+  def get_by_slug!(slug, opts \\ []) do
```

Comment:
@DouglasLutz / @maartenvanvliet what do you think about doing away with this function and instead amending `get!/1` like such:

```elixir
def get!(key, opts \\ []) do
  preloads = Keyword.get(opts, :preloads, [])

  query =
    from(c in Company)
    |> preload(^preloads)
    |> from()
    |> where([c], is_nil(c.removed_pending_change_id))

  final_query =
    case Integer.parse(key) do
      :error -> where(query, [c], c.slug == ^key)
      {int_id, _remainder} -> where([c], c.id == ^int_id or c.slug == ^key) 
    end

  Repo.one!(final_query)
end
```
While this function is a bit more complicated, I'm wondering if simplify things. We no longer need to change the param key to "slug" and update all occurrences of "id" accordingly. We handle both slug and id searches. All we need is the migration, schema + changeset, and this. :thinking: 

What do you both think?

Reply:
Yes, I had a local proof of concept with slugs that worked more along these lines. I see you account for slugs being numbers, I didn't :P 

Advantage is old integer based urls continue to work and a simpler internal api.

Reply:
> I see you account for slugs being numbers, I didn't :P

It hit me last night that there could realistically be a company who's name is just digits and that could throw a wrench in things.

While this code works I can't help but feel this is still "wrong". I _think_ we should be trying to identify when someone uses the old path (w/ ID) and 301 them to the new slug path. Thoughts?